### PR TITLE
don't treat UnavailableException as connection exception

### PR DESCRIPTION
**Pull request by jnelson15**

### CODE REVIEW

**Comment by schlosna** [comment]
We should probably add some tests describing the cases for how we're handling these exceptions.

**Comment by clockfort** [comment]
I don't think this will behave how we want, but I think this is partially because I tried very hard to make the code here very simple, and this is a problem with the simple way I handled host redirection during retries.

Looking at handleException, we're currently using isConnectionException to let us know we should blacklist or not. The act of blacklisting here has a side effect that, when we do the retry (up to n max retries) we will be choosing other random hosts to do the retry on rather than hammering the same one.

If this commit went through, we would get an UnavailableException and might try and hammer on that one bad node for all of our retry attempts, which is almost sure not to work.

If we're not going to instantly blacklist nodes with insufficient/unavailable exceptions, we need additional logic in runWithRetryOnHost to switch out the host we're talking after one or two tries on the original intended host.

tl;dr:

two options as I see it (some careful combinations of things here could also work):

-  still blacklist maybe-isolated nodes with unavailable exceptions, also make insufficient consistency a blacklistable exception, and make blacklist and pool refresh timers shorter than they are now; and deal with very logging-intensive death throws for the case when we actually can't continue because there aren't enough hosts up for real.

- make this change + change runWithRetryOnHost to understand that it should move to a different host after one or two tries on the intended host + make another change to blacklist isolated nodes by making a refresh-pool-time health check that examines all nodes' local system keyspace to see who hasn't recently gossip heartbeated to whom.

**Comment by jnelson15** [comment]
Yeah you're right -- currently it would retry on the same host which we don't want to do. Can we determine if an UnavailableException was thrown due to consistency errors at the client pool layer? If so, we could simply throw immediately and not retry since the retry would inevitably fail on any live host. Then wouldn't have to blacklist or worry about adding additional logic to runWithRetryOnHost to fail over to other live hosts.

**Comment by clockfort** [comment]
Unfortunately the gossip heartbeat thing I mentioned will be a trailing indicator for a cluster-wide consistency issue, (only able to detect the issue many seconds after it happens) and cannot be guaranteed on by logic here as a source of truth as to whether it's just this one guy who can't talk to other people or if everyone can't talk to each other with sufficient consistency.

**Comment by jnelson15** [comment]
Oh I see what you mean. Another alternative could be increasing the pool refresh rate when this happens. Downside of that is the live hosts will constantly be added/removed from the blacklist until we recover sufficient consistency. So not very ideal either.
